sirepo/job_driver/__init__.py

```diff
@@ -163,27 +187,23 @@ def make_lib_dir_symlink(self, op):
         """Only one op of each type allowed
 
         """
+        n = op.opName
+        if n in (job.OP_CANCEL, job.OP_KILL):
+            return
+        if n == job.OP_SBATCH_LOGIN:
+            l = [o for o in self.ops.values() if o.opId != op.opId]
+            assert not l, \
+                'received {} but have other ops={}'.format(op, l)
+            return
+        if op.op_slot:
+            return
+        q = self.op_q[n]
         try:
-            n = op.opName
-            if n in (job.OP_CANCEL, job.OP_KILL):
-                return
-            if n == job.OP_SBATCH_LOGIN:
-                l = [o for o in self.ops.values() if o.opId != op.opId]
-                assert not l, \
-                    'received {} but have other ops={}'.format(op, l)
-                return
-            if op.op_slot:
-                return
-            q = self.op_q[n]
-            try:
-                op.op_slot = q.get_nowait()
-            except tornado.queues.QueueEmpty:
-                pkdlog('self={} op={} await op_q.get()', self, op)
-                op.op_slot = await q.get()
-                raise job_supervisor.Awaited()
-        finally:
-            # may be redundant, but need to track here
-            self.ops[op.opId] = op
```

Comment:
The op would be lost if it exited due to an assertion  (line 195).

Reply:
The insertion of the op in self.ops was moved into `prepare_send`. `prepare_send` is the only caller of this function (I should make it private in that case) so I don't think it will be lost.